pkg/cloudevents/event_data.go

```diff
@@ -14,6 +14,38 @@ import (
 
 // SetData implements EventWriter.SetData
 func (e *Event) SetData(obj interface{}) error {
```

Comment:
I find SetData confusing. Its behavior depends on whether SetDataContentType was called beforehand, and the case of passing []byte when contenttype is set is ambiguous: am I asking for []byte to be further encoded in the content type (e.g. "base64" for JSON) or am I providing []byte that are *already* JSON encoded? The existing API can handle both cases but it depends on setting content-type and data in the right order. 

Here's an alternative API, with a backwards-compatible SetData() implementation:

```go
package cloudevents

import (
	"context"
	"mime"

	"github.com/cloudevents/sdk-go/pkg/cloudevents/datacodec"
	"github.com/cloudevents/sdk-go/pkg/cloudevents/types"
)

// ================ A safer API: WARNING UNTESTED CODE :)

// SetDataBytes sets DataContentType=contentType, and Data=bytes.
//
// bytes must already be encoded according contentType.
//
func (e *Event) SetDataBytes(contentType string, bytes []byte) { // No error!
	e.Context.SetDataContentType(contentType)
	e.Data = bytes
}

// MarshalDataBytes sets DataContentType=contentType, marshals v as contentType,
// and sets Data=marshaled-bytes.
//
// contentType must bed a content-type supported by package datacodec.
// For other content types, encode the data elsewhere and call SetDataBytes()
//
func (e *Event) MarshalDataBytes(contentType string, v interface{}) error {
	// FIXME(alanconway) should do mime parse in datacodec.Encode(),
	// not here. The datacodec will need the full content-type to handle
	// charset etc.
	mediaType, _, err := mime.ParseMediaType(contentType)
	if err != nil {
		return err
	}
	b, err := datacodec.Encode(context.Background(), mediaType, v)
	if err != nil {
		return err
	}
	e.SetDataBytes(contentType, b)
	return err
}

// SetDataValue clears DataContentType and sets Data to one of the basic
// CloudEvents data types:
//
//    bool, int32, string, []byte, *url.URL, time.Time
//
// Data is set to types.Validate(v), but not encoded.
// When the event itself is encoded, Data will be encoded according to the
// CloudEvents default encoding rules.
//
func (e *Event) SetDataValue(v interface{}) error {
	v, err := types.Validate(v)
	if err == nil {
		e.Context.SetDataContentType("")
		e.Data = v
	}
	return err
}

// ================ Backward compatible SetData:

// SetData encodes obj as the event Data according to
// Context.GetDataContentType()
//
// If the content-type is empty, obj must be convertible to one of the
// CloudEvents basic types, and will be default encoded when the event
// is marshaled. The basic types are:
//
//    bool, int32, string, []byte, *url.URL, time.Time
//
// Changing the content-type after calling SetData may create an invalid event.
//
func (e *Event) SetData(v interface{}) error {
	if contentType := e.Context.GetDataContentType(); contentType != "" {
		return e.SetDataValue(v)
	} else {
		return e.MarshalDataBytes(contentType, v)
	}
}

```

Reply:
A big change deserves big comments :)

Reply:
I agree. SetData no longer works as expected in 1.0 world.

https://github.com/cloudevents/sdk-go/issues/231

Reply:
I think I will attempt this in a followup with adding the NOP data encoder.